language/move-lang/src/shared/graph.rs

```diff
@@ -0,0 +1,360 @@
+                finger2 = self.idom_tree[&finger2];
+            }
+        }
+        finger1
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test1() {
+        let nodes = vec![1, 2, 3, 4, 5];
+        let edges = vec![(1, 2), (2, 1), (4, 1), (3, 2), (5, 3), (5, 4)];
+        let source = 5;
+        let graph = Graph::new(source, nodes, edges);
+        assert!(graph.compute_reducible().is_none());
+    }
+
+    #[test]
+    fn test2() {
+        let nodes = vec![1, 2, 3, 4, 5, 6];
+        let edges = vec![
+            (1, 2),
+            (2, 1),
+            (2, 3),
+            (3, 2),
+            (4, 2),
+            (4, 3),
+            (5, 1),
+            (6, 5),
+            (6, 4),
+        ];
+        let source = 6;
+        let graph = Graph::new(source, nodes, edges);
+        assert!(graph.compute_reducible().is_none());
+    }
+
+    #[test]
+    fn test3() {
+        let nodes = vec![1, 2, 3, 4, 5, 6];
+        let edges = vec![(1, 2), (2, 3), (2, 4), (2, 6), (3, 5), (4, 5), (5, 2)];
+        let source = 1;
+        let graph = Graph::new(source, nodes, edges);
+        let natural_loops = graph.compute_reducible().unwrap();
+        assert_eq!(natural_loops.len(), 1);
+        let single_loop = &natural_loops[0];
+        assert_eq!(single_loop.loop_header, 2);
+        assert_eq!(single_loop.loop_latch, 5);
+        assert_eq!(
+            single_loop.loop_body,
+            vec![2, 3, 4, 5].into_iter().collect()
+        );
+    }
+
+    #[test]
+    fn test4() {
+        let nodes = vec![1, 2, 3, 4, 5, 6];
+        let edges = vec![(1, 2), (2, 3), (2, 4), (2, 6), (3, 5), (4, 5), (5, 2)];
+        let source = 6;
+        let graph = Graph::new(source, nodes, edges);
+        assert!(graph.compute_reducible().is_some());
+    }
+
+    #[test]
+    fn test5() {
+        // nested natural loops
+        let nodes = vec![1, 2, 3, 4, 5, 6];
+        let edges = vec![
+            (1, 2),
+            (2, 3),
+            (2, 4),
+            (2, 6),
+            (3, 5),
+            (4, 5),
+            (5, 2),
+            (3, 2),
+        ];
+        let source = 1;
+        let graph = Graph::new(source, nodes, edges);
+        let natural_loops = graph.compute_reducible().unwrap();
+
+        assert_eq!(natural_loops.len(), 2);
+        let (inner_loop, outer_loop) = if natural_loops[0].loop_latch == 3 {
+            assert_eq!(natural_loops[1].loop_latch, 5);
+            (&natural_loops[0], &natural_loops[1])
+        } else {
+            assert_eq!(natural_loops[0].loop_latch, 5);
+            assert_eq!(natural_loops[1].loop_latch, 3);
+            (&natural_loops[1], &natural_loops[0])
+        };
+
+        assert_eq!(inner_loop.loop_header, 2);
+        assert_eq!(inner_loop.loop_body, vec![2, 3].into_iter().collect());
+
+        assert_eq!(outer_loop.loop_header, 2);
+        assert_eq!(outer_loop.loop_body, vec![2, 3, 4, 5].into_iter().collect());
+    }
+}
```

Comment:
This file is copied from #7952 which originally exists as `move-prover/bytecode/src/graph.rs`